src/kernel/arch/x86_64/cpu/apic.c

```diff
@@ -2,16 +2,97 @@
 #include <cpu.h>
 #include <video.h>
 #include <msr.h>
+#include <bitmap.h>
+#include <idt.h>
+#include <pic8259.h>
+
+extern uint32_t memory_size_in_bytes;
+uint32_t apic_base_address;
 
 void init_apic(){
     uint32_t apic_supported = _cpuid_feature_apic();
     if (apic_supported == 0x100){
-        _printStr("Apic supported\n");
+        printf("Apic supported\n");
     }
     uint64_t msr_output = rdmsr(IA32_APIC_BASE);
     printf("APIC MSR Return value: 0x%x\n", msr_output);
-    printf("APIC MSR Return value: 0x%x\n", (msr_output&APIC_BASE_ADDRESS_MASK));
-
+    printf("APIC MSR Base Address: 0x%x\n", (msr_output&APIC_BASE_ADDRESS_MASK));
+    apic_base_address = (msr_output&APIC_BASE_ADDRESS_MASK);
+    if(apic_base_address == NULL){
+        printf("ERROR: cannot determine apic base address\n");
+    }
+    map_phys_to_virt_addr(apic_base_address, apic_base_address, 0);
+    uint32_t *spurious_interrupt_register = (uint64_t *) (apic_base_address + APIC_SPURIOUS_VECTOR_REGISTER_OFFSET);
     printf("Apic enabled: %x\n", 1&(msr_output >> APIC_GLOBAL_ENABLE_BIT));
     printf("Apic BSP: %x\n", 1&(msr_output >> APIC_BSP_BIT));
+    
+    if(!(1&(msr_output >> APIC_GLOBAL_ENABLE_BIT))){
+        printf("Apic disabled globally");
+        return;
+    }
+    
+    if(!(1&(*spurious_interrupt_register >> APIC_SOFTWARE_ENABLE_BIT))){
+        printf("Apic disable via software...\nEnabling it\n");
+        //TODO enable apic software bit
```

Comment:
It would be worth implementing this before we merge, for completeness! In fact we probably dont need the check. You can just write to the spurious interrupt `vector | (1 << 8)`, everything else being set to zero is the default behavior.